### PR TITLE
Cleanup memory management in RegEx and File I/O

### PR DESCRIPTION
**Pull request by densh**
* Simplify cre2 bindings, don't try to hide low-level aspects under helper methods
* Make naming in cre2 to be more consistent with the original library
* Update `alloc` to zero-initialize the memory
* Make sure we don't GC allocate manually, this isn't supported on immix, use zones instead

### CODE REVIEW

File: nativelib/src/main/scala/scala/scalanative/native/package.scala [hunk 1]
```diff
@@ -147,11 +148,14 @@ package object native {
...
-  /** Convert a java.lang.String to a CString using given charset. */
+  /** Convert a java.lang.String to a CString using given charset.
+   *  The resulting c-style string is allocated using malloc, and
+   *  must be freed manually.
+   */
   def toCString(str: String,
                 charset: Charset = Charset.defaultCharset()): CString = {
     val bytes = str.getBytes(charset)
-    val cstr  = GC.malloc_atomic(bytes.length + 1).cast[Ptr[Byte]]
+    val cstr  = malloc(bytes.length + 1).cast[Ptr[Byte]]
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/native/package.scala]: @Duhemm This is a major change that we have to do to support precise garbage collection. I've seen that you use `GC.malloc` a lot in the File I/O, those have to be updated asap to do manual malloc/free. This also includes all results of `toCString` invocations.

File: javalib/src/main/scala/java/util/regex/Pattern.scala [hunk 2]
```diff
@@ -107,10 +106,12 @@ object Pattern {
...
 
   def quote(s: String): String = {
-    val original = StringPart(s)
-    val quoted   = StringPart.stackalloc
+    val original, quoted = stackalloc[cre2.string_t]
+    toRE2String(s, original)
     cre2.quoteMeta(quoted, original)
-    quoted.toString
+    val res = fromRE2String(quoted)
+    free(original.data.cast[Ptr[Byte]])
```
Comment (ekrich) [inline javalib/src/main/scala/java/util/regex/Pattern.scala]: You need to free a stack allocated String or should this be malloc as you are passing back into the Scala world? 
`free(original.data.cast[Ptr[Byte]])`

Comment (ekrich) [inline javalib/src/main/scala/java/util/regex/Pattern.scala]: Or the returned string needs to be managed by the runtime?

Comment (densh) [inline javalib/src/main/scala/java/util/regex/Pattern.scala]: `fromRE2String` calls `toCString`, which uses `malloc` to allocate the memory. It means we must free it afterwards.

Comment (ekrich) [inline javalib/src/main/scala/java/util/regex/Pattern.scala]: Thanks, will dig deeper next time.

File: nativelib/src/main/scala/scala/scalanative/native/package.scala (hunk 1 above)
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/native/package.scala]: Actually, doing `free` is horrendously error prone, lets use memory zones instead https://github.com/scala-native/scala-native/pull/734

I'll rebase regex PR on top once the zones are merged.

**Comment by densh** [comment]
@Duhemm Can you please review this one? It touches up File I/O quite a bit.

File: javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala [hunk 3]
```diff
@@ -34,71 +33,120 @@ final class NativePosixFileAttributeView(path: Path,
...
-        FileTime.from(!(sb._7), TimeUnit.SECONDS)
-      override val lastModifiedTime =
-        FileTime.from(!(sb._8), TimeUnit.SECONDS)
-      override val creationTime =
-        FileTime.from(!(sb._9), TimeUnit.SECONDS)
-
-      override val group = new GroupPrincipal {
-        override val getName = fromCString(!(_group._1))
+      private def fileStat()(implicit z: Zone) =
+        getStat()
```
Comment (Duhemm) [inline javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala]: Any reason why you want to re-retrieve the `stat` struct for each query on an attribute?

Comment (densh) [inline javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala]: If you cache unmanaged pointer in a managed object, then you'll need to clean it up only when object itself is unreachable (i.e. garbaged collected.) On JVM you can use finalizers for that, but we don't support them at the moment, hence no caching.

Comment (Duhemm) [inline javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala]: Thanks for the clarifications!

**Pull request merged**